File: jvm/src/main/kotlin/SlimFramework/logger.kt
package SlimFramework

import java.io.File
import java.time.LocalDateTime

//  use semaphore so threads don't fight over logging
private val s = FrwSemaphore(1, "Logger")
var GlobalLoggingLevel = LoggingLevel.none
var GlobalLoggingFile = "mylog.log"

//  for logging errors
fun error(classname: String, routine: String, msg: String) {
    if (GlobalLoggingLevel >= LoggingLevel.error) {
        writeLine("ERR", classname, routine, msg)
    }
}

//  for logging debug
fun debug(classname: String, routine: String, msg: String) {
    if (GlobalLoggingLevel >= LoggingLevel.debug) {
        writeLine("DBG", classname, routine, msg)
    }
}

//  for logging information
fun info(classname: String, routine: String, msg: String) {
    if (GlobalLoggingLevel >= LoggingLevel.info) {
        writeLine("INF", classname, routine, msg)
    }
}

//  write the logging information to the console and/or logfile
private fun writeLine(logtype: String, classname: String, routine: String, msg: String) {
    s.lock()
    val line = "TIME:${LocalDateTime.now()}::${logtype}::CLASS:${classname}::FUN:${routine}: ${msg}.\n"
    File(GlobalLoggingFile).appendText(line, Charsets.UTF_8)
    print(line)
    s.unlock()
}

//  enum for logging levels
enum class LoggingLevel {
    none,
    error,
    info,
    debug
}

File: dotnet/Logger.cs
using System;
using System.Threading;
using System.IO;

namespace Framework
{

    public static class Logger
    {

        static FrwSemaphore s = new FrwSemaphore(1, "FrwLogger");
        public static LoggingLevels GlobalLoggingLevel { get; set; } = LoggingLevels.None;
        const string path = "./mvc.log";

        public static void Error(string classname, string routine, string msg)
        {
            if (GlobalLoggingLevel >= LoggingLevels.Errors)
            {
                Log("ERR", classname, routine, msg);
            }
        }

        public static void Debug(string classname, string routine, string msg)
        {
            if (GlobalLoggingLevel >= LoggingLevels.Debug)
            {
                Log("DBG", classname, routine, msg);
            }

        }

        public static void Info(string classname, string routine, string msg)
        {
            if (GlobalLoggingLevel >= LoggingLevels.Info)
            {
                Log("INF", classname, routine, msg);
            }

        }

        public static void Extra(string classname, string routine, string msg)
        {
            if (GlobalLoggingLevel >= LoggingLevels.Everything)
            {
                Log("EXT", classname, routine, msg);
            }
        }

        static void Log(string prefix, string classname, string routine, string msg)
        {
            s.Lock();
            string line = $"{DateTime.Now.ToString()}::{classname}::{routine}::{prefix}: {msg}\n";
            Console.Write(line);
            using (var file = File.AppendText(path))
            {
                file.Write(line);
            }
            s.Unlock();
        }
    }
}